builder/docker_test.go

```diff
@@ -0,0 +1,170 @@
+package builder
+
+import (
+	// "fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"testing"
+)
+
+func TestGetDirFromConfigPath(t *testing.T) {
+	tests := []struct {
+		Filepath        string
+		ExpectedDirPath string
+		ShouldFail      bool
+	}{
+		{
+			Filepath:        "some/path/to/directory",
+			ExpectedDirPath: "some/path/to/directory",
+		},
+		{
+			Filepath:        "some/path/to/file",
+			ExpectedDirPath: "some/path/to",
+		},
+		{
+			Filepath:        "some/path/to/doesnotexist",
+			ExpectedDirPath: "some/path/to",
+		},
+
+		// Error cases.
+		{Filepath: "not/a/real/path", ShouldFail: true},
+	}
+
+	testDir, err := ioutil.TempDir("", "getdirfrompath-test-")
+	if err != nil {
+		t.Fatalf("couldn't create temporary directory to write test cases: %s", err)
+	}
+	defer func() {
+		_ = os.RemoveAll(testDir)
+	}()
+
+	// Set up the dirs/files for testing
+	path := filepath.Join(testDir, "some/path/to/directory")
+	if err = os.MkdirAll(path, 0777); err != nil {
+		t.Errorf("Failed to create test directory: %q\n%s", path, err)
+	}
+	path = filepath.Join(testDir, "some/path/to/file") // "some/path/to" already exists from previous command
+	f, err := os.Create(path)
+	if err != nil {
+		t.Errorf("Failed to create test file: %q\n%s", path, err)
+	}
+	defer func() {
+		_ = f.Close()
+	}()
+
+	for _, tt := range tests {
+		fullPath := filepath.Join(testDir, tt.Filepath)
+
+		dir, err := getDirFromConfigPath(fullPath)
+		dir = strings.TrimPrefix(dir, testDir+"/")
+		failed := err != nil
+		if failed != tt.ShouldFail {
+			if tt.ShouldFail {
+				t.Errorf("unexpected success when parsing filepath\nFILEPATH: %s\nPARSED DIR: %s\n", tt.Filepath, dir)
+			} else {
+				t.Errorf("unexpected error parsing bundle: %s\nFILEPATH: %s\nEXPECTED DIR: %s\n", err, tt.Filepath, tt.ExpectedDirPath)
+			}
+			continue
+		}
+		if tt.ShouldFail {
+			continue
+		}
+
+		if !reflect.DeepEqual(dir, tt.ExpectedDirPath) {
+			t.Errorf("got wrong dir when parsing filepath\nFILEPATH: %s\nPARSED DIR: %s\nEXPECTED DIR: %s", tt.Filepath, dir, tt.ExpectedDirPath)
+		}
+	}
+}
+
+func TestCanAccess(t *testing.T) {
+	tests := []struct {
+		Filepath   string
+		ShouldFail bool
+	}{
+		{Filepath: "shouldwork"},
+
+		// Error cases.
+		{Filepath: "noread", ShouldFail: true},
+		{Filepath: "nowrite", ShouldFail: true},
+		{Filepath: "noreadwrite", ShouldFail: true},
+		{Filepath: "/usr/var", ShouldFail: true},
```

Comment:
We expect this to fail because we expect the user to not have access to `/usr/var`, yet you are already testing this case with the `noreadwrite` test above, and not relying on expected test-runner permissions. I'd remove this case, then remove the `!strings.HasPrefix(fullPath, "/")` check below.

Reply:
True. I was originally thinking that it is very subtly a different failure, 'cause it's being denied from a group access failure, but it's not really proving anything new. I'll pull it.